src/compiled_proof.rs

```diff
@@ -0,0 +1,437 @@
+                ret.extend(&proof_bytes);
+            }
+            Command::Hash => {
+                ret.push(3u8);
+            }
+            Command::HashPeak => {
+                ret.push(4u8);
+            }
+            Command::ToPeak => {
+                ret.push(5u8);
+            }
+        }
+    }
+
+    Ok(ret)
+}
+
+pub fn pack_leaves<T: Packable>(leaves: &[(u64, T)]) -> Result<Vec<u8>> {
+    let mut ret = Vec::new();
+    for (pos, item) in leaves {
+        ret.extend(&pos.to_le_bytes()[..]);
+        ret.extend(item.pack()?);
+    }
+    Ok(ret)
+}
+
+pub struct PackedMerkleProof<'a, T> {
+    index: usize,
+    data: &'a [u8],
+    merge: PhantomData<T>,
+}
+
+impl<'a, T> PackedMerkleProof<'a, T> {
+    pub fn new(data: &'a [u8]) -> Self {
+        Self {
+            index: 0,
+            data,
+            merge: PhantomData,
+        }
+    }
+}
+
+impl<'a, T: Packable> Iterator for PackedMerkleProof<'a, T> {
+    type Item = Result<Command<T>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index >= self.data.len() {
+            return None;
+        }
+
+        let command = self.data[self.index];
+        self.index += 1;
+
+        match command {
+            1u8 => Some(Ok(Command::NextLeaf)),
+            2u8 => match T::unpack(&self.data[self.index..]) {
+                Ok((proof, size)) => {
+                    self.index += size;
+                    Some(Ok(Command::Proof(proof)))
+                }
+                Err(e) => Some(Err(e)),
+            },
+            3u8 => Some(Ok(Command::Hash)),
+            4u8 => Some(Ok(Command::HashPeak)),
+            5u8 => Some(Ok(Command::ToPeak)),
+            _ => Some(Err(Error::InvalidCommand(command))),
+        }
+    }
+}
+
+pub struct PackedLeaves<'a, T> {
+    index: usize,
+    data: &'a [u8],
+    t: PhantomData<T>,
+}
+
+impl<'a, T> PackedLeaves<'a, T> {
+    pub fn new(data: &'a [u8]) -> Self {
+        Self {
+            index: 0,
+            data,
+            t: PhantomData,
+        }
+    }
+}
+
+impl<'a, T: Packable> Iterator for PackedLeaves<'a, T> {
+    type Item = Result<(u64, T)>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index >= self.data.len() {
+            return None;
+        }
+
+        if self.data.len() - self.index < 8 {
+            return Some(Err(Error::UnpackEof));
+        }
+
+        let pos = {
+            let mut buf = [0u8; 8];
+            buf.copy_from_slice(&self.data[self.index..self.index + 8]);
+            u64::from_le_bytes(buf)
+        };
+        self.index += 8;
+
+        match T::unpack(&self.data[self.index..]) {
+            Ok((item, size)) => {
+                self.index += size;
+                Some(Ok((pos, item)))
+            }
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+enum StackItemType {
+    Node,
+    Peak,
+    Proof,
+}
+
+pub fn calculate_root<
+    T: PartialEq + Clone,
```

Comment:
unnecessary `PartialEq` trait bound, and we may remove the clone code in this fn and remove the trait bounds for T
```suggestion
    T,
```